File: sorting Algorithm/5. QuickSort.ts
function QuickSort(arr: number[], low: number, high: number) {
  if (low < high) {
    let pi: number = partition(arr, low, high);
    console.log('pi', pi);
    QuickSort(arr, low, pi - 1);
    QuickSort(arr, pi + 1, high);
  }

  return arr;
}

function partition(arr: number[], low: number, high: number) {
  let pivot = arr[low];
  let i = low;
  let j = high;
  while (i < j) {
    while (arr[i] <= pivot && i <= high - 1) {
      i++;
    }

    while (arr[j] > pivot && j >= low + 1) {
      j--;
    }

    if (i < j) {
      let temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    }
  }
  let temp = arr[j];
  arr[low] = temp;
  arr[j] = pivot;
  return j;
}

console.log(QuickSort([10, 80, 30, 90, 40, 50, 70], 0, 6));

File: sorting Algorithm/5. QuickSort.py
class Solution:
    def QuickSort(self, arr, low, high):
        if (low < high):
            pi = self.findPivot(arr, low, high)
            self.QuickSort(arr, low, pi-1)
            self.QuickSort(arr, pi+1, high)

    def findPivot(self, arr, low, high):
        pivot = arr[low]
        i = low
        j = high
        while i < j:
            while arr[i] <= pivot and i <= high:
                i += 1

            while arr[j] > pivot and j >= low:
                j -= 1

            if i < j:
                [arr[i], arr[j]] = [arr[j], arr[i]]

        [arr[low], arr[j]] = [arr[j], arr[low]]
        return j


arr = [4, 6, 2, 5, 7, 9, 1, 3]
s = Solution()
s.QuickSort(arr, 0, len(arr)-1)
print(arr)
